core/src/main/java/org/axonframework/serialization/compression/AbstractSerializerCompressionWrapper.java

```diff
@@ -0,0 +1,117 @@
+     */
+    public AbstractSerializerCompressionWrapper(Serializer embeddedSerializer, Converter converter) {
+        if(!embeddedSerializer.canSerializeTo(byte[].class)) {
+            throw new IllegalArgumentException("Expecting serializer that can serialize to byte array.");
+        }
+        this.embeddedSerializer = embeddedSerializer;
+        this.converter = converter;
+    }
+
+    @Override
+    public <T> SerializedObject<T> serialize(Object object, Class<T> expectedRepresentation) {
+        if(converter.canConvert(byte[].class, expectedRepresentation)) {
+            SerializedObject<byte[]> embeddedSerializedObject = embeddedSerializer.serialize(object, byte[].class);
+            try {
+                SerializedObject<byte[]> compressedSerializedObject = new SimpleSerializedObject<>(
+                        doCompress(embeddedSerializedObject.getData()),
+                        embeddedSerializedObject.getContentType(),
+                        embeddedSerializedObject.getType());
+                return converter.convert(compressedSerializedObject, expectedRepresentation);
+            } catch (Throwable e) {
+                throw new SerializationException("Unhandled Exception while compressing.", e);
+            }
+        } else {
+            throw new SerializationException("Converter unable to convert from byte[] after compressing.");
+        }
+    }
+
+    @Override
+    public <T> boolean canSerializeTo(Class<T> expectedRepresentation) {
+        return converter.canConvert(byte[].class, expectedRepresentation);
+    }
+
+    @Override
+    public <S, T> T deserialize(SerializedObject<S> serializedObject) {
+        if(converter.canConvert(serializedObject.getContentType(), byte[].class)) {
+            try {
+                SerializedObject<byte[]> convertedSerializedObject = converter.convert(serializedObject, byte[].class);
+                SerializedObject<byte[]> decompressedSerializedObject = new SimpleSerializedObject<>(
+                        doDecompress (convertedSerializedObject.getData()),
+                        convertedSerializedObject.getContentType(),
+                        convertedSerializedObject.getType()
+                );
+                return embeddedSerializer.deserialize(decompressedSerializedObject);
+            } catch (Throwable e)  {
+                throw new SerializationException("Unhandled Exception while decompressing.", e);
+            }
+        } else {
+            throw new SerializationException("Converter unable to convert to byte[] prior to decompressing.");
+        }
+    }
+
+    @Override
+    public Class classForType(SerializedType type) throws UnknownSerializedTypeException {
+        return embeddedSerializer.classForType(type);
+    }
+
+    @Override
+    public SerializedType typeForClass(Class type) {
+        return embeddedSerializer.typeForClass(type);
+    }
+
+    @Override
+    public Converter getConverter() {
```

Comment:
I see some issues when using this serializer in combination with Upcasters. I think, with a minor modification, this wrapper can be made upcaster friendly.
In fact, the Converter also needs to be wrapped. Whenever a byte[] is provided as a source, this converter should decompress it, before passing the data to the wrapped converter. This (converter) wrapper instance should also be capable of dealing with byte[] that doesn't contain compressed data.